src/Mvc/Mvc.Formatters.Json/src/TranscodingWriteStream.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+using System;
+using System.Buffers;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Mvc.Formatters.Json
+{
+    internal sealed class TranscodingWriteStream : Stream
+    {
+        private readonly Stream _stream;
+        private readonly ArrayPool<char> _charPool;
+        private readonly Encoding _targetEncoding;
+
+        public TranscodingWriteStream(Stream stream, Encoding targetEncoding)
+        {
+            _stream = stream;
+            _charPool = ArrayPool<char>.Shared;
+            _targetEncoding = targetEncoding;
+        }
+
+        public override bool CanRead => true;
+        public override bool CanSeek => false;
+        public override bool CanWrite => false;
+        public override long Length => throw new NotSupportedException();
+        public override long Position { get; set; }
+
+        public override void Flush()
+            => throw new NotSupportedException();
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            return _stream.FlushAsync(cancellationToken);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        private void ThrowArgumentOutOfRangeException(byte[] buffer, int offset, int count)
+        {
+            if (buffer.Length - offset < count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ThrowArgumentOutOfRangeException(buffer, offset, count);
+
+            var maxChars = Encoding.UTF8.GetMaxCharCount(count);
```

Comment:
I'm not sure about this. An alternative implementation would be to use the encoder and just pick a buffer size. What if maxChars is huge?

Reply:
Yeah, I need to clean this up. Likely used a fixed size char buffer

Reply:
if maxchars is huge it will be pooled?

Reply:
The array pool will return a heap allocated buffer if the size if too big or the pool is exhausted.